compiler+runtime/src/cpp/jank/read/parse.cpp

```diff
@@ -365,56 +359,100 @@ namespace jank::read::parse
         obj::persistent_hash_map::create_unique(std::make_pair(splicing_allowed_var, jank_true)))
       .expect_ok();
     util::scope_exit const finally{ [] { __rt_ctx->pop_thread_bindings().expect_ok(); } };
+    native_vector<processor::object_result> const items(begin(), end());
 
-    native_unordered_map<runtime::object_ref, object_source_info> parsed_keys{};
-    /* TODO: Only use an array map if everything can fit. */
-    runtime::detail::native_persistent_array_map ret;
-    for(auto it(begin()); it != end(); ++it)
+    if(expected_closer.is_some())
     {
-      if(it.latest.unwrap().is_err())
-      {
-        return err(it.latest.unwrap().expect_err());
-      }
-      auto const key(it.latest.unwrap().expect_ok().unwrap());
+      return error::parse_unterminated_map(start_token.start);
+    }
 
-      if(++it == end())
-      {
-        return error::parse_odd_entries_in_map({ start_token.start, latest_token.end },
-                                               { key.start.start, key.end.end });
-      }
+    expected_closer = prev_expected_closer;
 
-      if(it.latest.unwrap().is_err())
+    native_unordered_map<runtime::object_ref, object_source_info> parsed_keys{};
+
+    auto const build_map([&](auto &ret) -> processor::object_result {
```

Comment:
`auto &ret` isn't actually being returned, so perhaps `auto &map` is a better name.